enhancements/storage/csi-driver-install.md

```diff
@@ -79,61 +79,187 @@ along with in-tree driver and users could use both.
+
+Each CSI driver is deployed via a dedicated operator (i.e. each CSI driver has its own), to accommodate differences
+between the CSI drivers. For example, Manila CSI driver operator can detect if Manila service is present in underlying
+OpenStack cloud and deploy the driver only when the service is present.
+
+* Each CSI driver will provide its own CRD, `<backend>Driver` in `csi.openshift.io` API group. E.g. `AWSEBSDriver` or
+  `ManilaDriver`. The CRD is non-namespaced - CSI drivers are global in the whole cluster.
+* Each CRD will allow only a single instance, named `cluster`.
+* Initially, the CRDs will be equal and provide only common
+  [`OperatorSpec`](https://github.com/openshift/api/blob/95abe2d2f4223d5931e418bf8e4d3773d16b42c0/operator/v1/types.go#L48)
+  and [`OperatorStatus`](https://github.com/openshift/api/blob/95abe2d2f4223d5931e418bf8e4d3773d16b42c0/operator/v1/types.go#L97).
+  In the future, each CSI driver may introduce its own configuration fields.
+* Each CSI driver runs in a dedicated namespace, typically `openshift-<backed>-csi-driver` (`openshift-aws-ebs-csi-driver`).
+  * TODO: or run everything in cluster-storage namespace? OLM installs into openshift-aws-ebs-csi-driver, we would need to delete it before starting a new driver in cluster-storage. 
+* TODO: where the operator runs? cluster-storage namespace?
+* Each CSI driver uses cloud-credential-operator to obtain a role in the underlying cloud + its credentials to
+  manipulate with the cloud storage API.
+
+### Installation
+1. During installation, CVO starts cluster-storage-operator (CSO).
+2. CSO detects underlying cloud and starts operators for CSI driver(s) for the particular cloud.
+   * I.e. CSO deploys corresponding CSI driver operators, incl. their RBAC, CRD, Deployment and finally a default CR.
+3. The CSI driver operator obtains cluster credentials via cloud-credential-operator and deploys the CSI driver,
+   i.e. creates its RBAC, Deployment, DaemonSet and CSIDriver objects. Progress of the deployment is reported in
+   `CR.Status` of the operator CR.
+4. CSO monitors CRs of the CSI drivers and reports back their status ("Progressing", "Degraded", "Available", ...) in
+   CSO's ClusterOperator status.
+
+### Upgrade from OLM-managed operator
+In OCP 4.5, we released AWS EBS and Manila CSI drivers via OLM. It's too late to move them to CSO.
+We designed their operators in a way that regardless where the CSI driver operator runs, the driver itself always runs
+in openshift-aws-ebs-csi-driver namespace / openshift-manila-csi-driver.
+
+1. During upgrade, CVO starts new 4.6 cluster-storage-operator (CSO).
+2. 4.6 CSO detects that there is AWS / Manila driver installed by OLM.
+3. CSO deletes Subscription of the operator. OLM removes the operator, but leaves the CRD, CR and the driver / operand running.
+4. CSO runs the new CSI driver operator.
+5. The CSI driver operator adopts the old operand - it reconciles the old CSI driver objects (Deployment, DaemonSet,
```

Comment:
As far as I understand these operators are going out as tech previews. We do not need to make the tech previews upgrade to the GA. Instead of adopting and upgrading the tech preview functionality, we should consider documenting that the administrator needs to uninstall the tech preview before upgrading to 4.6. A few things to consider though if we went down that route:

1. Uninstalling the OLM operator (deleting the Subscription) won't cascade to uninstall the operand or anything else it was managing, in this case things like the StorageClass. What are the potential risks if things like the StorageClass and running drivers aren't cleaned up correctly and the CSO installed operator isn't able to adopt them?

2. What are the implications of asking the administrator to completely uninstall the drivers before upgrading? Will that result in losing data in PVs that had been backed by that driver? Do the PVs get deleted? Would we be orphaning the EBS volumes in some way?

Another tool we have are upgrade preflight checks. If we needed to we could write a preflight that makes sure the operator, APIs, etc don't exist on the cluster and prevent the upgrade until the administrator removes it. 

We will also be able to monitor how many installations of these tech preview operators exist through telemetry once they are shipped, this would help us weigh the decision of whether it is worth implementing a preflight, or doing the adoption of the tech preview.

Reply:
Unfortunately, Manila wants to be GA in 4.5. We already missed 4.4 where they wanted to be TP.
cc @Fedosin